retworkx-core/src/planar/lr_planar.rs

```diff
@@ -0,0 +1,699 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use std::cmp::Ordering;
+use std::hash::Hash;
+use std::vec::IntoIter;
+
+use hashbrown::{hash_map::Entry, HashMap};
+use petgraph::{
+    visit::{
+        EdgeCount, EdgeRef, GraphBase, GraphProp, IntoEdges, IntoNodeIdentifiers, NodeCount,
+        Visitable,
+    },
+    Undirected,
+};
+
+use crate::traversal::{depth_first_search, DfsEvent};
+
+type Edge<G> = (<G as GraphBase>::NodeId, <G as GraphBase>::NodeId);
+
+fn insert_or_min<K, V>(xs: &mut HashMap<K, V>, key: K, val: V)
+where
+    K: Hash + Eq,
+    V: Ord + Copy,
+{
+    xs.entry(key)
+        .and_modify(|e| {
+            if val < *e {
+                *e = val;
+            }
+        })
+        .or_insert(val);
+}
+
+fn edges_filtered_and_sorted_by<G, P, F, K>(
+    graph: G,
+    a: G::NodeId,
+    filter: P,
+    compare: F,
+) -> IntoIter<Edge<G>>
+where
+    G: IntoEdges,
+    P: Fn(&Edge<G>) -> bool,
+    F: Fn(&Edge<G>) -> K,
+    K: Ord,
+{
+    let mut edges = graph
+        .edges(a)
+        .filter_map(|edge| {
+            let e = (edge.source(), edge.target());
+            if filter(&e) {
+                Some(e)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+    edges.sort_by_key(compare);
+    // Remove parallel edges since they do *not* affect whether a graph is planar.
+    edges.dedup();
+    edges.into_iter()
+}
+
+fn is_target<G: GraphBase>(edge: Option<&Edge<G>>, v: G::NodeId) -> Option<&Edge<G>> {
+    edge.filter(|e| e.1 == v)
+}
+
+#[derive(Clone, Copy, PartialEq, PartialOrd)]
+struct Interval<T> {
+    inner: Option<(T, T)>,
+}
+
+impl<T> Default for Interval<T> {
+    fn default() -> Self {
+        Interval { inner: None }
+    }
+}
+
+impl<T> Interval<T> {
+    fn new(low: T, high: T) -> Self {
+        Interval {
+            inner: Some((low, high)),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.inner.is_none()
+    }
+
+    fn unwrap(self) -> (T, T) {
```

Comment:
Is there a reason this is not `&self`?

Reply:
Yeah, this function mirrors and calls [`Option::unwrap`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap) that consumes `self`.